### ISSUE TITLE
assume_role() in certain circumstances can break deterministic instance ids

### ISSUE THREAD

**Issue by mwmix** [getmoto/moto#6592]
Moto Version: 4.1.14
Issue:
Stopping all running mocks doesn't seem to completely reset moto. As a result unique identifiers can vary between runs. This is particularly frustrating in the context of unit testing. Where I can run all the tests individually and get one set of values, but once I try to run them sequentially I get another set of values. This is despite setting the seed and as far as I know telling moto to clean everything up.

The following minimal example demonstrates the issue I found when two unit tests were run back to back. If you run the code as is you will see the assert at the bottom fail. However, if you remove the "Unit Test 1" section and run the code it will pass. This doesn't seem to make any intuitive sense since we are "stop()"ing the relevant mocks and creating them anew between each test. (Original code was using python unittest but I got rid of that to hopefully simplify / minimize the code for you all)
```python
#!/usr/bin/python3
import moto
import boto3
import requests


mock_sts = None
mock_ec2 = None


def setUp():
    mock_sts = moto.mock_sts()
    mock_sts.start()

    mock_ec2 = moto.mock_ec2()
    mock_ec2.start()

    requests.post("http://motoapi.amazonaws.com/moto-api/seed?a=42")

    sts_client = boto3.client("sts")

    session_client = sts_client.assume_role(
        RoleArn="arn:aws:iam::111111111111:role/my-role",
        RoleSessionName="role-session-name"
    )

    session = boto3.session.Session(
        aws_access_key_id=session_client["Credentials"]["AccessKeyId"],
        aws_secret_access_key=session_client["Credentials"]["SecretAccessKey"],
        aws_session_token=session_client["Credentials"]["SessionToken"]
    )

    return mock_sts, mock_ec2, session


def tearDown():
    mock_sts.stop()
    mock_ec2.stop()

# Begin: Unit Test 1
mock_sts, mock_ec2, session = setUp()
sts_client = boto3.client("sts")

sts_client.assume_role(
    RoleArn="arn:aws:iam::111111111111:role/my-role",
    RoleSessionName="role-session-name"
)

mock_sts.stop()
mock_ec2.stop()
# End: Unit Test 1

# Begin: Unit Test 2
mock_sts, mock_ec2, session = setUp()
ec2_client = session.client("ec2")
instances = ec2_client.run_instances(
    MaxCount=1,
    MinCount=1
)["Instances"]

instance_ids = [instance["InstanceId"] for instance in instances]

print(f"instance_ids: {instance_ids}")

expected = ["i-9bd53fc993d07650f"]
print(f"expected: {expected}")

assert (instance_ids == expected)

mock_sts.stop()
mock_ec2.stop()
# End: Unit Test 2
```

What you see when you just run the above code as is:
```bash
./my_test2.py
instance_ids: ['i-2c04eb577967c8182']
expected: ['i-9bd53fc993d07650f']
Traceback (most recent call last):
  File "./my_test2.py", line 66, in <module>
    assert (instance_ids == expected)
AssertionError
```

What you see when you remove "Unit Test 1":
```bash
./my_test2.py
instance_ids: ['i-9bd53fc993d07650f']
expected: ['i-9bd53fc993d07650f']
```

I would of expected that it shouldn't matter if "Unit Test 1" was run or not since I was cleaning up after myself with each run.

Am I just missing some other stop() somewhere? Some sort of reset moto entirely kind of thing?

**Comment by bblommers** [comment]
Hi @mwmix! Thanks for the repro - that made it very easy to see the problem.

I think this behaviour is somehow related to the fact that Moto generates some static data when calling `start()`. When `test_2` is run on it's own, the `start()` doesn't know about the second account, `111111111111`.
If `test_1` is run first, by the time `test_2` executes `start()` it does know about the second account, so it will create the static data for it. And that additional data would explain the difference in 'unique' identifiers. 

I'm not quite sure what a fix for this would look like, to be honest. But as a workaround, you could create the second account explicitly before seeding:

```
from moto.iam import iam_backends
iam_backends["111111111111"]

requests.post("http://motoapi.amazonaws.com/moto-api/seed?a=42")
```
That would keep the instance ID's in `test2` persistent.

**Comment by bblommers** [comment]
Hi @mwmix and @adammiller1152, this should be fixed in the upcoming v5 release.

We've just released an alpha version to gather some early feedback, `moto==5.0.0alpha1`. Are you able to verify that this version fixes your issue?

Note that there are breaking changes in this release: all service-specific decorators have been replaced with a single `@mock_aws` decorator.
See the full release announcement here: https://github.com/getmoto/moto/issues/7198
